### PR TITLE
Adding MINGW32 toolchain support

### PR DESCRIPTION
**Pull request by Hombre57**

### CODE REVIEW

**Comment by clanmills** [comment]
Thanks for looking into this.  The last time I looked at MinGW/CMake, I had trouble building CMake and had to modify a Microsoft SDK file to even get it to compile.

I'll have a look at this tomorrow.  I'll be very interested in any comments you wish to make about getting CMake to work on MinGW.

**Comment by Hombre57** [comment]
@clanmills Today's versions of Mingw's package are quite stable. Cmake is provided as a Mingw package, no need to build it. I already have some packages installed for luminanceHDR, RawTherapee and HRDMerge, so I can't really say which one has to be installed for exiv2. If you need to know, I can have a look in the next few days.

**Comment by clanmills** [comment]
Thanks for the update.  I'll look at that on Tuesday.  I hope this works well as this is useful with Qt/MinGW.  I supported this for v0.26 and built it with the autotools.  For v0.27, we only want to support the CMake build and it would be great to get this operational again.

**Comment by Hombre57** [comment]
@clanmills I managed to build everything using Cmake only with this simple patch and is used by digiKam installed here.

**Comment by clanmills** [comment]
I'm unable to make progress with this.   When you say:
```
Cmake is provided as a Mingw package, no need to build it.
```
How/where can I get this package?

I've spent 1.5 hours on this.  Can't build CMake from source (needs a C++11 compiler).  mingw-get g++ is 6.3 and won't compile anything at all (not even hello world).  Very frustrating.  I always find it very difficult to get anything to work on MinGW.

**Comment by Hombre57** [comment]
@clanmills Well, I'm using Migw through [MSYS2](http://www.msys2.org/) on Windows, which provide ready to use [packages](https://github.com/msys2/msys2/wiki/Packages) (this list is probably obsolete). All I have to do to install the toolchain is :

```
pacman -S mingw-w64-x86_64-gcc mingw-w64-x86_64-gdb mingw-w64-x86_64-make mingw-w64-x86_64-pkg-config mingw-w64-x86_64-cmake
```

See [the complete document](http://rawpedia.rawtherapee.com/Windows) for our Rawtherapee project.

So maybe I was a bit confusing and it doesn't work as smoothly on Linux, but it works fine with MSYS2 on Windows.

**Comment by clanmills** [comment]
Thanks.

Somebody spoke to me about building with MSYS2 last year (one of the German guys on darktable, I believe).  He said it was building, although  I couldn't get it to work.  Thanks.  CMake is working.

There's a problem with libintl which I can disable with `$ cmake .. -DEXIV2_ENABLE_NLS=0`, however it's still looking for #include <libintl.h>

CMake says `Failed to find Gettext libintl (missing: Intl_INCLUDE_DIR)`

There are hundreds of identical messages from CMake about:
```
System is unknown to cmake, create:
Platform/MINGW32_NT-10.0 to use this system, please send your config file ....
```

I'm wondering if we're mixing mingw32 and 64 as I have the file /mingw64/include/libintl.h

**Comment by Hombre57** [comment]
@clanmills Please tell me more about the system your building on.

It is also important to open the correct MSYS console (but I guess you saw that already [in first paragraph](http://rawpedia.rawtherapee.com/Windows)) : `MSYS` for the package management, and `MSYS32` or `MSYS64` for configuring and building your application.

I'm not at home, so can't check now, but let me know if you need more help, and I'll check in 7-8h.

**Comment by Hombre57** [comment]
@clanmills

...and have you installed the `gettext` package ?

```
$ pacman -S mingw-w64-x86_64-gettext
```

**Comment by clanmills** [comment]
Thanks for the tips.  Good News.  I know what's wrong!  My build is picking up an obsolete version of exv_conf.h.  I've pulled down a fresh copy of the code and it compiles.

---------------

Here's my conversation with Tobias about M2 last year.  http://dev.exiv2.org/boards/3/topics/2913

Which shell should I use as there are three ways to start M2 bash shell:
1) MSYS2
```
250 rmills@rmillsmm-w7:~ $ uname -a
MSYS_NT-10.0 rmillsmm-w7 2.9.0(0.318/5/3) 2017-09-13 23:16 x86_64 Msys
251 rmills@rmillsmm-w7:~ $
```
It fails to link as follows.  I know what that's about and can fix it later.
```
CMakeFiles/exiv2lib.dir/version.cpp.o:version.cpp:(.text+0x72a): undefined reference to `EnumProcessModules'
```

2) Mingw-w64 32 bit
```
251 rmills@rmillsmm-w7:~ $ uname -a
MINGW32_NT-10.0 rmillsmm-w7 2.9.0(0.318/5/3) 2017-09-13 23:16 x86_64 Msys
252 rmills@rmillsmm-w7:~ $ gettext --version
gettext (GNU gettext-runtime) 0.19.8.1
Copyright (C) 1995-1997, 2000-2007 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Written by Ulrich Drepper.
253 rmills@rmillsmm-w7:~ $
```

Thousands of linker errors.  I suspect it's generated a .so instead of a .dll.lib

3) Mingw-w64 64 bit (I haven't tried this).
```
252 rmills@rmillsmm-w7:~ $ uname -a
MINGW64_NT-10.0 rmillsmm-w7 2.9.0(0.318/5/3) 2017-09-13 23:16 x86_64 Msys
253 rmills@rmillsmm-w7:~ $
```

**Comment by Hombre57** [comment]
@clanmills I'm in France and it's +1h from you, yes we'll find a solution to you problem tomorrow.

So you're on Win10, I hope that's not the problem. Here is my system info

**MINGW32**
```
Jean-Christophe@MyPC MINGW32 ~
$ uname -a
MINGW32_NT-6.1 MyPC 2.10.0(0.325/5/3) 2018-02-09 15:25 x86_64 Msys
Jean-Christophe@MyPC MINGW32 ~
$ gettext --version
gettext (GNU gettext-runtime) 0.19.8.1
Copyright (C) 1995-1997, 2000-2007 Free Software Foundation, Inc.
Licence GPLv3+ : GNU GPL version 3 ou ultérieure <http://gnu.org/licenses/gpl.html>
Ceci est un logiciel libre : vous pouvez le modifier et le redistribuer.
Il n'y a PAS DE GARANTIE, dans la mesure de ce que permet la loi.
Écrit par Ulrich Drepper.
Jean-Christophe@MyPC MINGW32 ~
$
```

**MINGW64**
```
Jean-Christophe@MyPC MINGW64 ~
$ uname -a
MINGW64_NT-6.1 MyPC 2.10.0(0.325/5/3) 2018-02-09 15:25 x86_64 Msys
Jean-Christophe@MyPC MINGW64 ~
```

You should probably install the last version of MSYS2. Either install [the one provided on their website](http://www.msys2.org/), or update through an `MSYS` console using [section III of this page](https://github.com/msys2/msys2/wiki/MSYS2-installation).

**Comment by Hombre57** [comment]
... then you should use `MSYS64` to configure and build, since your system is 64 bits. Don't ask me for cross-compiling though.

**Comment by clanmills** [comment]
I'm lost.  I reinstalling the latest MSYS 32 and 64 on Windows 10.

```
512 rmills@rmillsmm:~ $ dir ~/Downloads/
-rw-r--r--+  1 rmills staff  76M Jun 27 09:40 /Users/rmills/Downloads/msys2-x86_64-20180531.exe
-rw-r--r--+  1 rmills staff  77M Jun 27 09:38 /Users/rmills/Downloads/msys2-i686-20180531.exe
```

There are a myriad of combinations and I can't get any of them to build!

We can select from 3 shells (in both 32 and 64)
1) MSYS
2) MinGW64
3) MinGW32

We can choose MSYS Makefiles on MinGW Makefiles:
```
$ cmake --help | grep -ie msys -ie ming
  MSYS Makefiles               = Generates MSYS makefiles.
  MinGW Makefiles              = Generates a make file for use with
                                 mingw32-make.
```
I assume you've successfully built with your patch.  Can you tell me exactly how so that I can repeat your build.

**Comment by clanmills** [comment]
I'm going to focus on the MinGW-64 bash shell.   Excellent documentation here:  http://rawpedia.rawtherapee.com/Windows

```
$ uname -a
MINGW64_NT-10.0 rmillsmm-w7 2.10.0(0.325/5/3) 2018-02-09 15:25 x86_64 Msys
```

```
pacman -S tar gzip nano make diffutils intltool git
```

```
pacman -S mingw-w64-x86_64-gcc mingw-w64-x86_64-gdb mingw-w64-x86_64-make mingw-w64-x86_64-pkg-config mingw-w64-x86_64-cmake
```
I'll update you later.

**Review by clanmills — COMMENTED** [review]
Great News.  I've got this built and running the test suite.  There are numerous test suite failures, however I'm not going to be distracted by that because we're re-engineering the test suite in python for v0.27.

Can you "untarget" CYGWIN for REGX_LIBRRARIES in src/CMakeLists.txt#262-268.  For example:

```
    if (CYGWIN OR MINGW)
        target_link_libraries( exiv2lib PRIVATE  psapi ws2_32 )
    endif()
    
    if (MINGW)
        target_link_libraries( exiv2lib PRIVATE ${REGX_LIBRARIES} )
    endif()    
```

**Comment by Hombre57** [comment]
@clanmills Great that you managed to link !

I added the patch you asked for in commit 9e84fa2.

Out of curiosity, does an Exiv2 IRC exist ? I didn't found any info on that.

**Comment by clanmills** [comment]
Team Exiv2 uses Slack for informal discussion with each other.  If you'd like to join us, please send your email address to robin@clanmills.com.  We regard Slack conversations as "off the record" and use GitHub for technical discussion.

I'll ask Luis (our CMake Expert) to review, approve and merge your changes.  His life is rather busy at the moment, so expect a delay before this is submitted to 'master'.

I have successfully built with 32 bit MinGW.  Very pleased that we have this working again.  Thank You for working with me.  I'll update the build notes appropriately.

File: cmake/findDependencies.cmake [hunk 1]
```diff
@@ -10,6 +10,14 @@ endif()
 
 find_package(Threads REQUIRED)
 
+if(MINGW)
+    find_package(PkgConfig REQUIRED)
+    pkg_check_modules(REGX REQUIRED regex)
```
Comment (piponazo) [inline cmake/findDependencies.cmake]: [minor] Could you please change `REGX` to `REGEX`? With just one additional character we can improve the readability of the code 😉

**Review by piponazo — APPROVED** [review]
Thanks for the contribution! Everything looks fine from the CMake point of view. I am just wondering if this code would be also needed for the linux distributions, but until now we never had problems with that.

File: cmake/findDependencies.cmake (hunk 1 above)
Comment (Hombre57) [reply cmake/findDependencies.cmake]: @piponazo `REGEX` has been recognized as a keyword by Notepad++, so I didn't even tried to use it... Fixed in commit 815a686

**Comment by piponazo** [comment]
I see that your new commit implements the change that I have requested, but this commit has not been pushed to this PR yet. Would you mind to squash both commits and update the PR? I will merge the changes as soon as it is done.

**Comment by Hombre57** [comment]
@piponazo ...which proves that I'm really not comfortable with forks :disappointed: I'll do that when back home.

**Comment by D4N** [comment]
Hey folks, I have tried to cross-compile with mingw using this patch but I can't get over the configure stage, as the cmake config won't pick find a regex library. Do we want to investigate that further or not since this PR seems to be targeted from MSYS2 only?

**Comment by piponazo** [comment]
@D4N with the version in master, are you able to configure & compile the project? This would mean that your MinGW version has internally regex.

**Comment by piponazo** [comment]
By the way, AppVeyor provides in their environment MinGW, MSYS, Cygwin:

https://www.appveyor.com/docs/build-environment/#mingw-msys-cygwin

In the future we could think of adding jobs to CI to check if we can at least, configure and compile the project in those platforms.

**Comment by Hombre57** [comment]
@piponazo I managed to add a commit with the remaining stuff in it, so the `mingw-support` branch is complete now. I didn't squashed everything because I still need to learn how Git works, but I guess it shouldn't harm (?)

**Comment by piponazo** [comment]
Thank @Hombre57 for the effort. I would like to test it my by self and maybe add a new CI job for MinGW before merging this. According to what @D4N found out, the condition `if (MINGW)` might not be needed for `MINGW`  but for cygwin.

I'll keep you updated with my findings.

**Comment by clanmills** [comment]
Thanks, @piponazo 

1) It's possible that the REGEX code also works on Cygwin.  I'll try to find time today to look at that.  2) I'd like to update README.md with notes about using MinGW+msys/2.
I'll send you a modified file.  Perhaps you could review/submit for me to GitHub.
3) My inclination is keep MinGW+msys/1.0 dead for reasons documented in TODO-CMAKE
The v0.26 public builds include MinGW+msys/1.0 which was created using autotools.

**Comment by piponazo** [comment]
@D4N What do you think about changing the condition `if(MINGW)` to `if(MINGW AND WIN32)` ? In that way it would not affect to the cross-compilation from Linux, and still, we could merge this contribution.

**Comment by clanmills** [comment]
What is your definition of "cross-compiling"?  Are you thinking about using Linux to build for MinGW?  Andreas used to do that, although I've never got it to work.

I looked at Cygwin+REGEX.  I can't get Cmake+Cygwin+GCC 6.4 to build anything.  I can build on the laptop with Cygwin+GCC 4.9, however I haven't tried the REGEX code on that machine.

Shridar has been keeping busy with Visual Studio and EPS (yuck).  Steady progress - a bit like England in the World Cup really!

**Comment by D4N** [comment]
> What is your definition of "cross-compiling"? Are you thinking about using Linux to build for MinGW? 

That's exactly what I tried.

@piponazo I am not sure if it's worth adding that in this PR. The cross compilation build is broken anyway, so if we want to support that, we'll have to fix it in a separate PR.

**Comment by clanmills** [comment]
**_Opinion:_**
Let's try to support and document cross compiling.

**_Related issue:_**
I've fixed an issue with src/getopt_win32 on Cygwin: https://github.com/Exiv2/exiv2/issues/377

**_REGX + CYGWIN doesn't work:_**:
I tried REGX on Cygwin with CMake 3.9.0 and GCC 4.9 (which builds/tests OK).

I've added the OR CYGWIN to cmake/findDependancies.cmake#13
```
if(MINGW OR CYGWIN)
    find_package(PkgConfig REQUIRED)
    pkg_check_modules(REGX REQUIRED regex)
    if(REGX_FOUND)
        include_directories(${REGX_INCLUDE_DIRS})
    endif()
endif()
```
``` 
529 rmills@rmillsmbp-w7:~/gnu/github/exiv2/exiv2/build $ cmake ..
-- Found PkgConfig: /usr/bin/pkg-config.exe (found version "0.29.1")
-- Checking for module 'regex'
--   No package 'regex' found
CMake Error at /usr/local/share/cmake-3.9.0/Modules/FindPkgConfig.cmake:412 (message):
  A required package was not found
Call Stack (most recent call first):
  /usr/local/share/cmake-3.9.0/Modules/FindPkgConfig.cmake:588 (_pkg_check_modules_internal)
  cmake/findDependencies.cmake:15 (pkg_check_modules)
  CMakeLists.txt:89 (include)

**Comment by piponazo** [comment]
Let's merge this (since our CI jobs are green) and in case we need to do add some additional logic in the future we can do it in a different PR.

**Comment by clanmills** [comment]
Thanks, Luis.  I'm going to work on that later today.  I'll have changes (documentation for sure) which I'll put into a different PR.  Everything's coming together well.

**Pull request merged**